addon/PreviewProvider.js

```diff
@@ -359,15 +359,34 @@ PreviewProvider.prototype = {
   },
 
   /**
-   * Do some pre-processing on the links before inserting them into the metadata
+   * Do some pre-processing on the link before inserting it into the metadata
    * DB and adding them to the metadata cache
    */
-  processAndInsertMetadata(links) {
-    const processedLinks = this._processLinks(links);
-    this.insertMetadata(processedLinks);
+  processAndInsertMetadata(link, metadataSource) {
+    const processedLink = this._processLinks(link);
+    this.insertMetadata(processedLink, metadataSource);
   },
 
   /**
+   * Check if a single link exists in the metadata DB
+   */
+  asyncDoesSingleLinkExist: Task.async(function*(url) {
+    let link = this._processLinks([{url}])[0];
+    if (!link) {
+      return false;
+    }
+    // Hit the cache first
+    const key = link.cache_key;
+    const metadataFromCache = MetadataCache.cache.get(key);
+
+    if (!metadataFromCache) {
+      const linksMetadata = yield this._metadataStore.asyncExecuteQuery(`SELECT * FROM page_metadata WHERE cache_key = '${key}'`);
```

Comment:
```asyncExecuteQuery``` was exposed so that we could make custom queries - I didn't want to use ```asyncGetMetadataByCacheKey``` because it expects an array of cache keys and does a WHERE IN query + a bunch of other work (like calling ```asynExecuteQuery``` twice to grab images and whatnot), which is unnecessary if you have just a single cache key look up and just need to check if it exists in the DB or not